monop/monop.6.in
.\"	$NetBSD: monop.6,v 1.13 2002/09/26 18:32:03 wiz Exp $
.\"
.\" Copyright (c) 1980 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)monop.6	6.5 (Berkeley) 3/25/93
.\"
.Dd March 25, 1993
.Dt MONOP 6
.Os
.Sh NAME
.Nm monop
.Nd Monopoly game
.Sh SYNOPSIS
.Nm
.Op Ar file
.Sh DESCRIPTION
.Nm
is reminiscent of the Parker Brother's game Monopoly, and
monitors a game between 1 to 9 users.
It is assumed that the rules of Monopoly are known.
The game follows the standard rules, with the exception that,
if a property goes up for auction and there are only two solvent players,
no auction is held and the property remains unowned.
.Nm
The game, in effect, lends the player money,
so it is possible to buy something which you cannot afford.
However, as soon as a person goes into debt, he must
.Dq fix the problem ,
.Em i.e. ,
make himself solvent, before play can continue.
If this is not possible, the player's property reverts to his debtee,
either a player or the bank.
A player can resign at any time to any person or the bank,
which puts the property back on the board, unowned.
.Pp
Any time that the response to a question is a
.Em string ,
e.g., a name, place or person, you can type
.Sq Ic ?
to get a list of valid answers.
It is not possible to input a negative number, nor is it ever necessary.
.Pp
.Em A Summary of Commands :
.Bl -tag -width item
.It Ic quit
quit game: This allows you to quit the game.
It asks you if you're sure.
.It Ic print
print board: This prints out the current board.
The columns have the following meanings (column headings are the same for the
.Ic where ,
.Ic own holdings  ,
and
.Ic holdings
commands):
.Pp
.Bl -tag -width indent -compact
.It Name
The first ten characters of the name of the square
.It Own
The
.Em number
of the owner of the property.
.It Price
The cost of the property (if any)
.It Mg
This field has a
.Sq *
in it if the property is mortgaged
.It \&#
If the property is a Utility or Railroad, this is the number
of such owned by the owner.
If the property is land, this is the number of houses on it.
.It Rent
Current rent on the property.
If it is not owned, there is no rent.
.El
.It Ic where
where players are: Tells you where all the players are.
A
.Sq *
indicates the current player.
.It Ic own holdings
List your own holdings,
.Em i.e. ,
money, get-out-of-jail-free cards, and property.
.It Ic holdings
holdings list: Look at anyone's holdings.
It will ask you whose holdings you wish to look at.
When you are finished, type
.Sq Ic done .
.It Ic mortgage
mortgage property:
Sets up a list of mortgageable property, and asks which you wish to mortgage.
.It Ic unmortgage
unmortgage property:
Unmortgage mortgaged property.
.It Ic buy
buy houses:
Sets up a list of monopolies on which you can buy houses.
If there is more than one, it asks you which you want to buy for.
It then asks you how many for each piece of property,
giving the current amount in parentheses after the property name.
If you build in an unbalanced manner
(a disparity of more than one house within the same monopoly),
it asks you to re-input things.
.It Ic sell
sell houses:
Sets up a list of monopolies from which you can sell houses.
It operates in an analogous manner to
.Ic buy .
.It Ic card
card for jail:
Use a get-out-of-jail-free card to get out of jail.
If you're not in jail, or you don't have one, it tells you so.
.It Ic pay
pay for jail:
Pay $50 to get out of jail, from whence you are put on Just Visiting.
Difficult to do if you're not there.
.It Ic trade
This allows you to trade with another player.
It asks you whom you wish to trade with,
and then asks you what each wishes to give up.
You can get a summary at the end, and, in all cases,
it asks for confirmation of the trade before doing it.
.It Ic resign
Resign to another player or the bank.
If you resign to the bank, all property reverts to its virgin state,
and get-out-of-jail free cards revert to the deck.
.It Ic save
save game:
Save the current game in a file for later play.
You can continue play after saving,
either by adding the file in which you saved the game after the
.Nm
command, or by using the
.Ic restore
command (see below).
It will ask you which file you wish to save it in,
and, if the file exists, confirm that you wish to overwrite it.
.It Ic restore
restore game:
Read in a previously saved game from a file.
It leaves the file intact.
.It Ic roll
Roll the dice and move forward to your new location.
If you simply hit the
.Aq RETURN
key instead of a command, it is the same as typing
.Ic roll .
.El
.Sh AUTHOR
.An Ken Arnold
.Sh FILES
.Bl -tag -width @monop_cardsfile@ -compact
.It Pa @monop_cardsfile@
Chance and Community Chest cards
.El
.Sh BUGS
No command can be given an argument instead of a response to a query.


NEWS
NEWS for bsd-games and bsd-games-non-free
=========================================

This file contains a summary of changes in each version of bsd-games
and bsd-games-non-free after 1.3.  For fuller details, see ChangeLog
and ChangeLog.0.

As well as reading this summary, it is a good idea to reread the
INSTALL file whenever installing a new version, to check for any
changes in installation instructions.  Packagers should also reread
the PACKAGING file each version (or do a diff between the old and new
versions of the package and read that).

Version 2.10 (2000-02-12)
============

* Avoid world-readable characs file in phantasia.  Note the new
configuration question which will need an appropriate answer,
especially if packaging for a distribution.

* Various minor bugfixes.

* Essentially all relevant changes, except to hunt (which will
eventually be replaced by the OpenBSD version), now merged back into
NetBSD.

* Some minor changes from NetBSD CVS of 2000-02-12.

* No new bsd-games-non-free version; bsd-games-non-free 2.8 is still
current.

Version 2.9 (1999-11-12)
===========

* Fixes various minor bugs reported recently.

* Some minor changes from NetBSD CVS of 1999-11-12.

* No new bsd-games-non-free version; bsd-games-non-free 2.8 is still
current.

Version 2.8 (1999-09-26)
===========

* Robots scorefile format has changed - remove or rename your old
score file when upgrading to this version.

* Various minor changes and bug-fixes (some from OpenBSD).

* Snake substantially cleaned up, and now uses the native curses
interface (rather than termcap emulation in ncurses); backgammon and
tetris still to be converted.

* New `game': wtf.

* Updated from NetBSD CVS of 1999-09-25; many more changes merged back
into NetBSD.

* All constant data files are now architecture independent (installed
under /usr/share by default).

* bsd-games is no longer tested by me with libc older than glibc
2.1.2; fixes for compilation with older libc will still be accepted.

* Details of the copyrights and licence terms applicable to these
games, as best as I can determine them, now appear in COPYING.

* bsd-games no longer comes with its own word list for boggle and
hangman; you need to provide one at compile time (for boggle) and at
run time (for hangman).  The copyright status of the one formerly
included was not entirely clear.  The word list used by hangman can
also now be specified at run time with the new `-d' option.

* bsd-games now has a testsuite (for programs that can be tested
non-interactively only).

Version 2.7 (1999-04-04)
===========

* Bug fixes.

* Updated from NetBSD-current of 1999-03-26; more changes merged back
into NetBSD.

* Other minor changes.

Version 2.6 (1998-12-20)
===========

* Buffer overrun fixes in sail (not thought exploitable).

* Fixes possible score file corruption if fds 0, 1, 2 aren't open at
game startup.

* Updated from NetBSD-current of 1998-12-07 (morse now has a decoding
option).

* Now builds cleanly with -Wwrite-strings.

* Other minor changes.

* Some users of this package may be interested in my `nppt' package,
which provides an alternative implementation of ppt(6) with additional
features such as Encapsulated PostScript output.  (This package
installs the binary as `nppt', so can be installed simultaneously with
this one.  Look for nppt-0.1.tar.gz (or a later version) on
metalab.unc.edu (formerly sunsite) in /pub/Linux/Incoming/ or under
/pub/Linux/games/.)

Version 2.5 (1998-09-24)
===========

* New build system, using a single project Makefile (autogenerated)
that includes makefile fragments from subdirectories, and uses GNU
Make's automatic dependency generation to generate dependencies
automatically, rather than recursive make.  See Peter Miller's paper
`Recursive Make Considered Harmful'
<http://www.canb.auug.org.au/~millerp/rmch/recu-make-cons-harm.html>
for reasons why this is better than recursive make.  This should make
no difference to normal users, but if you modify the sources note that
you will need to rebuild from the top directory, and you may find that
the dependencies are now accurate whereas formerly they were
incomplete.  The automatic dependency generation causes the build to
start with about 300 `no such file or directory' warnings, which
should be ignored: these refer to the dependency files which GNU Make
will then automatically generate.

* Should work with glibc 2.1 (at least, version 2.0.96 on x86, which
is what I've tested).  See the ChangeLog for an explanation of the
problem previous releases had with glibc 2.1.

* Boggle is now in the main bsd-games package - Paul Janzen checked
with the original author, who confirmed that the BSD licence
superseded the old non-free one.  bsd-games-non-free now only contains
rogue.

* Other minor changes.

Version 2.4 (1998-09-05)
===========

* Various minor bug fixes and improvements (some from OpenBSD, thanks
to Paul Janzen <pjanzen@foatdi.harvard.edu>).

* Sail directory can now be non-world-accessible (mode 2770 root.games
or similar).  This may improve security, since it means users cannot
tamper with sync files their sail process creates.

* Rogue buffer overrun fix.  An audit is still needed.

* dm now ignores the environment variable TZ when deciding at what
times games can be played.  (This only affects you if you are using dm
to control at what times games can be played, and not then if you are
using libc 5.4.45 or 5.4.46 (which completely ignore TZ in setgid
programs rather than checking the validity of any filename given as
glibc does).)

* All games updated from NetBSD-current of 1998-08-30 (minor changes;
some more of my changes have been merged back into NetBSD).

Version 2.3 (1998-08-19)
===========

* A possible temporary file security bug in sail fixed.

* Countmail is back, with a licence (apparently the standard NetBSD
one), since it has one in NetBSD-current of 1998-08-19.

* All the other games updated from NetBSD-current of 1998-08-19 (minor
changes).


monop/monop.c
/*	$NetBSD: monop.c,v 1.12 2001/09/18 18:15:49 wiz Exp $	*/

/*
 * Copyright (c) 1980, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1980, 1993\n\
	The Regents of the University of California.  All rights reserved.\n");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)monop.c	8.1 (Berkeley) 5/31/93";
#else
__RCSID("$NetBSD: monop.c,v 1.12 2001/09/18 18:15:49 wiz Exp $");
#endif
#endif /* not lint */

#include <stdio.h>
#include <signal.h>
#include <stdlib.h>
#include <unistd.h>
#include "monop.def"

int main __P((int, char *[]));
static void getplayers __P((void));
static void init_players __P((void));
static void init_monops __P((void));
static void do_quit __P((int));

/*
 *	This program implements a monopoly game
 */
int
main(ac, av)
	int ac;
	char *av[];
{
	/* Revoke setgid privileges */
	setregid(getgid(), getgid());

	srand(getpid());
	if (ac > 1) {
		if (!rest_f(av[1]))
			restore();
	}
	else {
		getplayers();
		init_players();
		init_monops();
	}
	num_luck = sizeof lucky_mes / sizeof (char *);
	init_decks();
	signal(SIGINT, do_quit);
	for (;;) {
		printf("\n%s (%d) (cash $%d) on %s\n", cur_p->name, player + 1,
			cur_p->money, board[cur_p->loc].name);
		printturn();
		force_morg();
		execute(getinp("-- Command: ", comlist));
	}
}

/*ARGSUSED*/
static void
do_quit(n)
	int n __attribute__((__unused__));
{
	quit();
}

/*
 *	This routine gets the names of the players
 */
static void
getplayers()
{
	char *sp;
	int i, j;
	char buf[257];

blew_it:
	for (;;) {
		if ((num_play=get_int("How many players? ")) <= 0 ||
		    num_play > MAX_PL)
			printf("Sorry. Number must range from 1 to 9\n");
		else
			break;
	}
	cur_p = play = (PLAY *) calloc(num_play, sizeof (PLAY));
	if (play == NULL)
		err(1, NULL);
	for (i = 0; i < num_play; i++) {
over:
		printf("Player %d's name: ", i + 1);
		for (sp = buf; (*sp=getchar()) != '\n'; sp++)
			continue;
		if (sp == buf)
			goto over;
		*sp++ = '\0';
		name_list[i] = play[i].name = (char *)calloc(1, sp - buf);
		if (name_list[i] == NULL)
			err(1, NULL);
		strcpy(play[i].name, buf);
		play[i].money = 1500;
	}
	name_list[i++] = "done";
	name_list[i] = 0;
	for (i = 0; i < num_play; i++)
		for (j = i + 1; j < num_play; j++)
			if (strcasecmp(name_list[i], name_list[j]) == 0) {
				if (i != num_play - 1)
					printf("Hey!!! Some of those are "
					    "IDENTICAL!!  Let's try that "
					    "again....\n");
				else
					printf("\"done\" is a reserved word.  "
					    "Please try again\n");
				for (i = 0; i < num_play; i++)
					free(play[i].name);
				free(play);
				goto blew_it;
			}
}

/*
 *	This routine figures out who goes first
 */
static void
init_players()
{
	int i, rl, cur_max;
	bool over = 0;
	int max_pl = 0;

again:
	putchar('\n');
	for (cur_max = i = 0; i < num_play; i++) {
		printf("%s (%d) rolls %d\n", play[i].name, i+1, rl=roll(2, 6));
		if (rl > cur_max) {
			over = FALSE;
			cur_max = rl;
			max_pl = i;
		}
		else if (rl == cur_max)
			over++;
	}
	if (over) {
		printf("%d people rolled the same thing, so we'll try again\n",
		    over + 1);
		goto again;
	}
	player = max_pl;
	cur_p = &play[max_pl];
	printf("%s (%d) goes first\n", cur_p->name, max_pl + 1);
}

/*
 *	This routine initializes the monopoly structures.
 */
static void
init_monops() 
{
	MON *mp;
	int i;

	for (mp = mon; mp < &mon[N_MON]; mp++) {
		mp->name = mp->not_m;
		for (i = 0; i < mp->num_in; i++)
			mp->sq[i] = &board[mp->sqnums[i]];
	}
}
